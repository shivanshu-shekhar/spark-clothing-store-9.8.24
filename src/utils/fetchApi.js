export const fetchApi = (apiUrl) => {

  console.log(apiUrl);
  return fetch(apiUrl)
    .then((res) => {
      // if success
      console.log(res);

      if (res.status === 200) {
        return res.json();
      } else {
        throw new Error("Invalid Response");
      }
    })
    .catch((err) => {
      console.log(err);
    });
    
};
