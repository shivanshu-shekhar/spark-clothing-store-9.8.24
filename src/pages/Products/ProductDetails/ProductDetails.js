import React, { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import { fetchApi } from "../../../utils/fetchApi";
import ProductsFeedback from "../ProductsFeedback/ProductsFeedback";

const ProductDetails = () => {
  let { productId } = useParams();
  const [data, setData] = useState([]);
  const [modalShow, setModalShow] = useState(null);
  const [isError, setIsError] = useState(false);
  const [isLoading, setIsLoading] = useState(true);
  const [userList, setUserList] = useState([]);

  useEffect(() => {
    //fetching the data from the backend server for particular product.
    fetchApi(`http://localhost:5001/data?id=${productId}`)
      .then((result) => {
        //console.log(result[0]);
        setData(result[0]);
        setIsLoading(false);
      })
      .catch((err) => {
        setIsError(true);
        setIsLoading(false);
      });

    //fetching the user feedback from the backend server and displaying it on the page.
    fetchApi(`http://localhost:5001/feedback?productId=${productId}`)
      .then((resInJSON) => {
        // capturing the converted JSON res
        // console.log(resInJSON);
        if ((resInJSON.status !== 404)) {//changed
          setIsLoading(false);
          setIsError(false);
          setUserList(resInJSON);
        } else {
          setIsLoading(false);
          setIsError(true);
          setUserList([]);
        }
      })
      .catch((err) => {
        console.log(err);
        setIsLoading(false);
        setIsError(true);
        setUserList([]);
      });
  }, [productId,userList]);

  //Spinner for the loading
  if (isLoading) {
    return (
      <div className="spinner-border text-success mt-5" role="status"></div>
    );
  }

  //Error message
  if (isError) {
    return (
      <div className="alert alert-danger mt-5">
        Some Error Occured! Please Try again later.
      </div>
    );
  }

  return (
    //Products details part start here.
    <div className="mt-5">
      <h1>Product Details: {data.title}</h1>
      <div className="card">
        <img
          className="card-img-top"
          src={data.img}
          width="100"
          height="1000"
        />
        <a className="text-center mt-3" href={data.url}>
          <button type="button" class="btn btn-dark">
            Visit the Official Store: {data.store}
          </button>
        </a>
        <div className="card-body text-center">
          <h4 className="card-title">{data.category}</h4>
          <h5 className="card-text">{data.title}</h5>
          <p className="card-text">
            <strong>Rs</strong> {data.reducedPrice} only
          </p>
          <p className="card-text">{data.description}</p>
          <p className="card-text">{data.fit}</p>
          <h5 className="card-text">Ratings :{data.rating}</h5>
        </div>
        {/* Products details part ends here. */}
        {/* Feeback button to submit the feedback. */}
        <button
          className="btn btn-primary"
          onClick={() => setModalShow(data.id)}
        >
          Click Here To Submit The Feedback
        </button>

        <div className="row mt-4">
          <h5>
            <strong>FEEDBACK BY THE USERS:</strong>
          </h5>
          {/* The feedbacks from the user */}
          {userList?.map((user) => {
            return (
              <div className="col-md-12" key={user.id}>
                <div className="card">
                  <div className="card-body">
                    <p className="card-title">
                      {/* user name */}
                      <strong>User Name: {user.name}</strong>
                    </p>
                    <p className="card-text">
                      {/* Feedback by the user */}
                      Feedback Given By The User:{" "}
                      <strong>{user.comment}</strong>
                    </p>
                    <p className="card-text">
                      {/* The rating by the user */}
                      Rating Given By The User:{" "}
                      <strong>{user.rate} stars</strong>
                    </p>
                  </div>
                </div>
              </div>
            );
          })}
        </div>
      </div>
      {modalShow && (
        <ProductsFeedback
          show={modalShow}
          onHide={() => setModalShow(null)}
          productId={modalShow}
        />
      )}
    </div>
  );
};

export default ProductDetails;
