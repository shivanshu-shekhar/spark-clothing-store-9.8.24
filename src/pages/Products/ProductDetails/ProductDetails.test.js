import { render, screen } from "@testing-library/react";
import { HashRouter } from "react-router-dom";
import { fetchApi } from "../../../utils/fetchApi";
import renderer from "react-test-renderer";
import ProductDetails from "./ProductDetails";

jest.mock("../../../utils/fetchApi");

const feedbackReviews = [
  {
    comment: "Good",
    name: "sourav",
  },
];

const mockItems = [
  {
    title: "Nike Top",
    category: "women",
    originalPrice: "2999",
  },
];

describe("ProductDetails", () => {
  //Postive test specs
  it("[MOCKING]: fetches productId properly via API Call", async () => {
    fetchApi.mockResolvedValue(mockItems);
    render(
      <HashRouter>
        <ProductDetails />
      </HashRouter>
    );
    expect(await screen.findByText(/women/i)).toBeInTheDocument();
  });

  it("[MOCKING]: fetches product data properly via API Call", async () => {
    fetchApi.mockResolvedValue(feedbackReviews);
    render(
      <HashRouter>
        <ProductDetails />
      </HashRouter>
    );
    expect(await screen.findByText(/sourav/i)).toBeInTheDocument();
  });

  //Negative test specs
  it("[MOCKING]: renders error properly during API Call ", async () => {
    // preparing mock error obj
    const error = "Error occured";

    fetchApi.mockRejectedValue(error);
    render(
      <HashRouter>
        <ProductDetails />
      </HashRouter>
    );
    expect(
      await screen.findByText(/Some Error Occured! Please Try again later./i)
    ).toBeInTheDocument();
  });

  test("renders correctly", () => {
    const tree = renderer
      .create(
        <HashRouter>
          <ProductDetails />
        </HashRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
