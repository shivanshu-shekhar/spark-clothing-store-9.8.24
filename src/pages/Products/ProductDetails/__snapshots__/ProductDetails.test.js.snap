// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails renders correctly 1`] = `
<div
  className="spinner-border text-success mt-5"
  role="status"
/>
`;
